File: Unit_tests_python/EX-01-compare_directories/Bard.py
import os

def compare_directories(dir1, dir2):
   """Compares two directories and returns a detailed report of the comparison."""

   report = {
       "Passed": True,
       "Failed_Count": 0,
       "Failed_Tests": []
   }

   # Get files from both directories
   dir1_files = set(os.listdir(dir1))
   dir2_files = set(os.listdir(dir2))

   # Check for files existing in one directory but not the other
   missing_files = dir1_files.symmetric_difference(dir2_files)
   if missing_files:
       report["Passed"] = False
       report["Failed_Count"] += 1
       report["Failed_Tests"].append({
           "Failed_Files": list(missing_files),
           "Failure_Location": "Both"  # Files missing in both directories
       })

   # Compare files with the same name in both directories
   common_files = dir1_files.intersection(dir2_files)
   for filename in common_files:
       filepath1 = os.path.join(dir1, filename)
       filepath2 = os.path.join(dir2, filename)

       # Check for content differences
       if not filecmp.cmp(filepath1, filepath2, shallow=False):
           report["Passed"] = False
           report["Failed_Count"] += 1
           report["Failed_Tests"].append({
               "Failed_Files": [filename],
               "Failure_Location": "Both"  # Content difference
           })

       # Check for modification date differences (considering sub-second precision)
       mtime1 = os.path.getmtime(filepath1)
       mtime2 = os.path.getmtime(filepath2)
       if mtime1 != mtime2:
           report["Passed"] = False
           report["Failed_Count"] += 1
           failure_location = "dir1" if mtime1 < mtime2 else "dir2"
           report["Failed_Tests"].append({
               "Failed_Files": [filename],
               "Failure_Location": failure_location
           })

   return report

File: Unit_tests_lua/EX-01-compare_directories/Bard.lua
function compare_directories(dir1, dir2)
    local report = {
      Passed = true,
      Failed_Count = 0,
      Failed_Tests = {},
    }
  
    -- Check if both directories exist
    if not io.dir(dir1) or not io.dir(dir2) then
      report.Passed = false
      report.Failed_Count = report.Failed_Count + 1
      table.insert(report.Failed_Tests, {
        Failed_Files = {"Both directories"},
        Failure_Location = "N/A",
      })
      return report
    end
  
    -- Helper function to compare file details
    local function compare_files(file1, file2)
      local stat1 = io.stat(dir1 .. "/" .. file1)
      local stat2 = io.stat(dir2 .. "/" .. file2)
  
      -- Check if files exist
      if not stat1 or not stat2 then
        report.Passed = false
        report.Failed_Count = report.Failed_Count + 1
        table.insert(report.Failed_Tests, {
          Failed_Files = {file1},
          Failure_Location = stat1 and "dir2" or "dir1",
        })
        return
      end
  
      -- Check file content
      if not io.open(dir1 .. "/" .. file1, "rb") or not io.open(dir2 .. "/" .. file2, "rb") then
        report.Passed = false
        report.Failed_Count = report.Failed_Count + 1
        table.insert(report.Failed_Tests, {
          Failed_Files = {file1},
          Failure_Location = "Error reading file content",
        })
        return
      end
      local content1 = io.read("*a")
      local content2 = io.read("*a")
      io.close()
      if content1 ~= content2 then
        report.Passed = false
        report.Failed_Count = report.Failed_Count + 1
        table.insert(report.Failed_Tests, {
          Failed_Files = {file1},
          Failure_Location = "Content mismatch",
        })
      end
  
      -- Check file modification dates
      if stat1.mtime ~= stat2.mtime then
        local failure_location = stat1.mtime < stat2.mtime and "dir1" or "dir2"
        report.Passed = false
        report.Failed_Count = report.Failed_Count + 1
        table.insert(report.Failed_Tests, {
          Failed_Files = {file1},
          Failure_Location = failure_location .. " has older modification date",
        })
      end
    end
  
    -- Iterate over files in dir1
    for filename in io.dir(dir1) do
      if filename ~= "." and filename ~= ".." then
        compare_files(filename, filename)
      end
    end
  
    return report
  end  